server/tilesource/base.py

```diff
@@ -352,29 +576,218 @@ def getRegion(self, width=None, height=None, **kwargs):
+        ratios = [round(ratio, 4) for ratio in ratios]
+        if not len(ratios):
+            return None if exact else mag['level']
+        if exact:
+            if any(int(ratio) != ratio or ratio != ratios[0]
+                   for ratio in ratios):
+                return None
+        ratio = round(sum(ratios) / len(ratios), 4)
+        level = mag['level'] + ratio
+        if rounding:
+            level = int(math.ceil(level) if rounding == 'ceil' else
+                        round(level))
+        if (exact and (level > mag['level'] or level < 0) or
+                (rounding == 'ceil' and level > mag['level'])):
+            return None
+        if rounding is not None:
+            level = max(0, min(mag['level'], level))
+        return level
+
+    def tileIterator(self, format=(TILE_FORMAT_PIL, ), resample=False,
+                     **kwargs):
+        """
+        Iterate on all tiles in the specifed region at the specified scale.
+        Each tile is returned as part of a dictionary that includes
+            x, y: (left, top) coordinates in current magnification pixels
+            width, height: size of current tile in pixels
+            tile: cropped tile image
+            format: format of the tile
+            level: level of the current tile. None if not present
+            level_x, level_y: the tile reference number within the level.
+            magnification: magnification of the current tile
+            mm_x, mm_y: size of the current tile pixel in millimeters.
+            gx, gy: (left, top) coordinate in maximum-resolution pixels
+            gwidth, gheight: size of of the current tile in maximum resolution
+                pixels.
+        If a region that includes partial tiles is requested, those tiles are
+        cropped appropriately.  Most images will have tiles that get cropped
+        along the right and bottom egdes in any case.  If an exact
+        magnification or scale is requested, no tiles will be returned unless
+        allowInterpolation is true.
+
+        :param format: the desired format or a tuple of allowed formats.
+            Formats are members of (TILE_FORMAT_PIL, TILE_FORMAT_NUMPY,
+            TILE_FORMAT_IMAGE).  If TILE_FORMAT_IMAGE, encoding must be
+            specified.
+        :param resample: If True or one of PIL.Image.NEAREST, LANCZOS,
+            BILINEAR, or BICUBIC to resample tiles that are not the target
+            output size.  Tiles that are resampled may have non-integer x, y,
+            width, and height values, and will have an additional dictionary
+            entries of:
+                scaled: the scaling factor that was applied (less than 1 is
+                    downsampled).
+                tile_x, tile_y: (left, top) coordinates before scaling
+                tile_width, tile_height: size of the current tile before
+                    scaling.
+            Note that scipy.misc.imresize uses PIL internally.
```

Comment:
Since you are returning this, i feel we should avoid exposing the `regionWidth` and `regionHeight` parameters. There are too many width and height parameters now that make things confusing.


*